From 148947fe867c094badfba970a55974c41177ff73 Mon Sep 17 00:00:00 2001
From: mj <mj@maxlinear.com>
Date: Thu, 14 Aug 2025 14:51:20 +0530
Subject: [PATCH] WLANRTSYS-92336 ACS should not remain on the same channel
 when selecting the second-best channel from the ACS table.

Issue: During ACS recalculation after radar detection, the same channel is being reselected.
Fix: To resolve this, the "stay on channel" logic is skipped when selecting
the second-best channel during ACS recalculation.
---
 src/ap/mxl_acs_proprietary.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 0a64454de..842df3dc5 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -3756,13 +3756,17 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 
 	}
 
-	if ((!iface->conf->mxl_conf.acs_bg_scan_do_switch && (SWR_BG_SCAN == switch_reason)) ||
-	    iface->mxl_iface.cac_required || !(iface->conf->mxl_conf.acs_init_done)) {
-		goto stay_on_chan;
-	}
+	if (!iface->mxl_iface.sec_best_ch.flag) {
+		if ((!iface->conf->mxl_conf.acs_bg_scan_do_switch && (SWR_BG_SCAN == switch_reason)) ||
+		    iface->mxl_iface.cac_required || !(iface->conf->mxl_conf.acs_init_done)) {
+			wpa_printf(MSG_DEBUG, "ACS: stay on channel reason SWR_BG_SCAN switch not enabled or cac required or acs init not done");
+			goto stay_on_chan;
+		}
 
-	if (!iface->conf->mxl_conf.acs_update_do_switch && (SWR_UPDATE == switch_reason)) {
-		goto stay_on_chan;
+		if (!iface->conf->mxl_conf.acs_update_do_switch && (SWR_UPDATE == switch_reason)) {
+			wpa_printf(MSG_DEBUG, "ACS: stay on channel reason SWR_UPDATE switch not enabled");
+			goto stay_on_chan;
+		}
 	}
 
 	for (prio = 0; prio < ACS_NUM_GRP_PRIORITIES; prio++) {
-- 
2.43.0

